File: src/teste.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX 500

typedef struct {
    int x, y;
} Point;

typedef struct {
    int u, v;
} Edge;

int N, M;
char maze[MAX][MAX];
int flood_gate[4];
int covers[MAX][2];
int num_covers;
Point path[MAX * MAX];
int path_len;

int dfn[MAX][MAX], low[MAX][MAX], idx, stack[MAX * MAX][2], top;
int bridges[MAX * MAX][4], bridge_count;
int visited[MAX][MAX];
int parent[MAX][MAX][2];

void add_bridge(int x1, int y1, int x2, int y2) {
    bridges[bridge_count][0] = x1;
    bridges[bridge_count][1] = y1;
    bridges[bridge_count][2] = x2;
    bridges[bridge_count][3] = y2;
    bridge_count++;
}

void tarjan_iterative(int sx, int sy) {
    memset(dfn, 0, sizeof(dfn));
    memset(low, 0, sizeof(low));
    memset(parent, -1, sizeof(parent));
    idx = 1;
    top = 0;

    stack[top][0] = sx;
    stack[top++][1] = sy;
    dfn[sx][sy] = low[sx][sy] = idx++;

    while (top > 0) {
        int x = stack[top-1][0];
        int y = stack[top-1][1];

        int found = 0;
        int adj[4][2] = {{x+1, y}, {x-1, y}, {x, y+1}, {x, y-1}};

        for (int i = 0; i < 4; i++) {
            int nx = adj[i][0], ny = adj[i][1];
            if (nx >= 0 && ny >= 0 && nx < N && ny < M && maze[nx][ny] != '#' && parent[x][y][0] != nx && parent[x][y][1] != ny) {
                if (!dfn[nx][ny]) {
                    stack[top][0] = nx;
                    stack[top++][1] = ny;
                    dfn[nx][ny] = low[nx][ny] = idx++;
                    parent[nx][ny][0] = x;
                    parent[nx][ny][1] = y;
                    found = 1;
                    break;
                } else {
                    low[x][y] = low[x][y] < dfn[nx][ny] ? low[x][y] : dfn[nx][ny];
                }
            }
        }

        if (!found) {
            top--;
            for (int i = 0; i < 4; i++) {
                int nx = adj[i][0], ny = adj[i][1];
                if (nx >= 0 && ny >= 0 && nx < N && ny < M && maze[nx][ny] != '#' && parent[x][y][0] != nx && parent[x][y][1] != ny) {
                    if (low[nx][ny] > dfn[x][y]) {
                        add_bridge(x, y, nx, ny);
                    }
                    low[x][y] = low[x][y] < low[nx][ny] ? low[x][y] : low[nx][ny];
                }
            }
        }
    }
}

void bfs_safe_path(int sx, int sy, int ex, int ey) {
    int queue[MAX * MAX][2], front = 0, rear = 0;
    int directions[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    memset(visited, 0, sizeof(visited));
    memset(parent, -1, sizeof(parent));
    visited[sx][sy] = 1;
    queue[rear][0] = sx;
    queue[rear++][1] = sy;

    while (front < rear) {
        int x = queue[front][0], y = queue[front][1];
        front++;

        if (x == ex && y == ey) break;

        for (int i = 0; i < 4; i++) {
            int nx = x + directions[i][0], ny = y + directions[i][1];
            if (nx >= 0 && ny >= 0 && nx < N && ny < M && maze[nx][ny] != '#' && !visited[nx][ny]) {
                visited[nx][ny] = 1;
                parent[nx][ny][0] = x;
                parent[nx][ny][1] = y;
                queue[rear][0] = nx;
                queue[rear++][1] = ny;
            }
        }
    }

    int cx = ex, cy = ey;
    path_len = 0;

    while (cx != sx || cy != sy) {
        path[path_len].x = cx;
        path[path_len].y = cy;
        int temp_x = parent[cx][cy][0];
        int temp_y = parent[cx][cy][1];
        cx = temp_x;
        cy = temp_y;
        path_len++;
    }
    path[path_len].x = sx;
    path[path_len].y = sy;
    path_len++;
}

int main() {
    int t;
    scanf("%d", &t);

    while (t--) {
        scanf("%d %d", &N, &M);

        for (int i = 0; i < N; i++) {
            scanf("%s", maze[i]);
        }

        int C;
        scanf("%d", &C);

        int sx = 0, sy = 0, ex = 0, ey = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (maze[i][j] == 'D') {
                    sx = i;
                    sy = j;
                } else if (maze[i][j] == 'E') {
                    ex = i;
                    ey = j;
                }
            }
        }

        bridge_count = 0;
        tarjan_iterative(sx, sy);
        
        for (int i = 0; i < bridge_count; i++) {
            printf("%d %d %d %d\n", bridges[i][0], bridges[i][1], bridges[i][2], bridges[i][3]);
        }

        // Use the first bridge as the flood gate location
        if (bridge_count > 0) {
            flood_gate[0] = bridges[0][0];
            flood_gate[1] = bridges[0][1];
            flood_gate[2] = bridges[0][2];
            flood_gate[3] = bridges[0][3];
        } else {
            // Default to (0,0)-(0,1) if no bridges found, adjust as necessary
            flood_gate[0] = 0;
            flood_gate[1] = 0;
            flood_gate[2] = 0;
            flood_gate[3] = 1;
        }

        // Place manhole covers (example: just cover the first C manholes)
        num_covers = 0;
        for (int i = 0; i < N && num_covers < C; i++) {
            for (int j = 0; j < M && num_covers < C; j++) {
                if (maze[i][j] == 'M') {
                    covers[num_covers][0] = i;
                    covers[num_covers][1] = j;
                    num_covers++;
                }
            }
        }

        bfs_safe_path(sx, sy, ex, ey);

        // Output
        printf("%d %d %d %d\n", flood_gate[0], flood_gate[1], flood_gate[2], flood_gate[3]);
        printf("%d\n", num_covers);
        for (int i = 0; i < num_covers; i++) {
            printf("%d %d\n", covers[i][0], covers[i][1]);
        }
        printf("%d\n", path_len);
        for (int i = path_len - 1; i >= 0; i--) {
            printf("%d %d\n", path[i].x, path[i].y);
        }
    }

    return 0;
}

File: src/teste.py
from collections import deque
import random
from sys import stdin, stdout
from itertools import combinations

DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def readln():
    return stdin.readline().rstrip()

def outln(n):
    stdout.write(str(n) + '\n')

def dfs_bridges_iterative(grid, sx, sy, ex, ey):
    rows, columns = len(grid), len(grid[0])
    visited = [[False] * columns for _ in range(rows)]
    disc = [[float('inf')] * columns for _ in range(rows)]
    low = [[float('inf')] * columns for _ in range(rows)]
    parent = [[None] * columns for _ in range(rows)]
    bridges = []
    stack = []
    time = [0]

    for i in range(rows):
        for j in range(columns):
            if (grid[i][j] == 'M' or grid[i][j] == '.') and not visited[i][j]:
                stack.append((i, j, 0))

                while stack:
                    x, y, state = stack.pop()

                    if not state:
                        visited[x][y] = True
                        disc[x][y] = low[x][y] = time[0]
                        time[0] += 1
                        stack.append((x, y, 1))

                        for dx, dy in DIRECTIONS:
                            nx, ny = x + dx, y + dy
                            if 0 <= nx < rows and 0 <= ny < columns and (grid[nx][ny] == '.' or grid[nx][ny] == 'M'):
                                if not visited[nx][ny]:
                                    stack.append((nx, ny, 0))
                                    parent[nx][ny] = (x, y)
                                elif (nx, ny) != parent[x][y]:
                                    low[x][y] = min(low[x][y], disc[nx][ny])
                    else:
                        for dx, dy in DIRECTIONS:
                            nx, ny = x + dx, y + dy
                            if 0 <= nx < rows and 0 <= ny < columns and (grid[nx][ny] == '.' or grid[nx][ny] == 'M'):
                                if parent[nx][ny] == (x, y):
                                    low[x][y] = min(low[x][y], low[nx][ny])
                                    if low[nx][ny] > disc[x][y]:
                                        if (x, y) < (nx, ny) and valid_flood_gate(grid, rows, columns, sx, sy, ex, ey, ((x, y), (nx, ny))):
                                            bridges.append(((x, y), (nx, ny)))
                                            return bridges
                                        elif (x, y) >= (nx, ny) and valid_flood_gate(grid, rows, columns, sx, sy, ex, ey, ((nx, ny), (x, y))):
                                            bridges.append(((nx, ny), (x, y)))
                                            return bridges
                                        #bridges.append(((x, y), (nx, ny)))
                                elif (nx, ny) != parent[x][y]:
                                    low[x][y] = min(low[x][y], disc[nx][ny])

    return bridges

def bfs_safe_path(maze, N, M, sx, sy, ex, ey):
    queue = deque([(sx, sy)])
    visited = [[False] * M for _ in range(N)]
    visited[sx][sy] = True
    parent = [[None] * M for _ in range(N)]

    while queue:
        x, y = queue.popleft()
        if x == ex and y == ey:
            path = []
            while parent[x][y] is not None:
                path.append((x, y))
                x, y = parent[x][y]
            path.append((sx, sy))
            path.reverse()
            return path
        
        for dx, dy in DIRECTIONS:
            nx, ny = x + dx, y + dy
            if 0 <= nx < N and 0 <= ny < M and not visited[nx][ny] and maze[nx][ny] != '#':
                visited[nx][ny] = True
                parent[nx][ny] = (x, y)
                queue.append((nx, ny))
    return []

def evaluate_manhole_cover_effectiveness(maze, N, M, manholes, sx, sy, flood_gate, C):
    flood_effectiveness = {}

    base_flooded = bfs_flood_control(maze, manholes, flood_gate)

    for comb in combinations(manholes, C):
        flood = bfs_flood_control(maze, set(manholes) - set(comb), flood_gate, comb)

        filtered_comb = tuple(m for m in comb if not flood[m[0]][m[1]])        
        if len(filtered_comb) != len(comb):
            comb = filtered_comb
        
        flood_effectiveness[comb] = sum(1 for x in range(N) for y in range(M) if base_flooded[x][y] and not flood[x][y])

    best_combination = max(flood_effectiveness, key=flood_effectiveness.get)
    return list(best_combination)


memo = {}
def bfs_flood_control(grid, start_points, flood_gate, block=()):
    key = (tuple(start_points), flood_gate, tuple(block))
    if key in memo:
        return memo[key]
    
    #print("start_points", start_points)
    rows, columns = len(grid), len(grid[0])
    queue = deque(start_points)
    
    if block:
        for i in block:
            if i in queue:
                queue.remove(i)
    
    visited = [[False] * columns for _ in range(rows)]
    
    #print("queue", queue)
    #print("flood_gate", flood_gate)
    # Marcar a flood gate como visitada
    #visited[flood_gate[0][0]][flood_gate[0][1]] = True
    #visited[flood_gate[1][0]][flood_gate[1][1]] = True
    
    
#    #print("sadfsdf", start_points)
    #for x, y in start_points:
    #    visited[x][y] = True

    #print("block", block)
    #if block:
    #    for i in block:
            ##print(i)
    #        grid[i[0]][i[1]] = '.'
        #x, y = block
        #grid[x][y] = '.'
        #queue.po
        #visited[block[0]][block[1]] = False
        
        #if (x, y) in start_points:
        #    return visited
        
       
    while queue:
        x, y = queue.popleft()
        #print("x, y:", x, y)
        
        for dx, dy in DIRECTIONS:
            nx, ny = x + dx, y + dy
            if 0 <= nx < rows and 0 <= ny < columns and not visited[nx][ny] and grid[nx][ny] != '#':
                #print("nx, ny", nx, ny, "dx dy", dx, dy)
                #print("sd", subtract_tuples(flood_gate[1], flood_gate[0]), "fff", flood_gate[0], flood_gate[1])
                #print("diferenca", (dx, dy) == subtract_tuples(flood_gate[1], flood_gate[0]))
                if block:
                    if (dx, dy) == (0, -1):
                        if (nx, ny) == flood_gate[0] and (x, y) == flood_gate[1]:
                            #print("continunen", nx, ny, (nx, ny) == flood_gate[0], (nx, ny) == flood_gate[1])
                            continue
                    elif (dx, dy) == (0, 1):
                        if (nx, ny) == flood_gate[1] and (x, y) == flood_gate[0]:
                            #print("continunen", nx, ny, (nx, ny) == flood_gate[0], (nx, ny) == flood_gate[1])
                            continue
                    elif (dx, dy) == (1, 0):
                        if (nx, ny) == flood_gate[1] and (x, y) == flood_gate[0]:
                            #print("continunen", nx, ny, (nx, ny) == flood_gate[0], (nx, ny) == flood_gate[1])
                            continue
                    elif (dx, dy) == (-1, 0):
                        if (nx, ny) == flood_gate[0] and (x, y) == flood_gate[1]:
                            #print("continunen", nx, ny, (nx, ny) == flood_gate[0], (nx, ny) == flood_gate[1])
                            continue
                
                visited[nx][ny] = True
                queue.append((nx, ny))

    memo[key] = visited
    return visited


def select_effective_covers(manhole_effectiveness):
    best_combination = max(manhole_effectiveness, key=manhole_effectiveness.get)
    return list(best_combination)


def valid_flood_gate(maze, N, M, sx, sy, ex, ey, bridge):
    (bx1, by1), (bx2, by2) = bridge
    maze[bx1][by1] = '#'
    maze[bx2][by2] = '#'
    
    path_exists = bfs_safe_path(maze, N, M, sx, sy, ex, ey) != []

    maze[bx1][by1] = '.'
    maze[bx2][by2] = '.'
    
    return path_exists


def main():
    num_cases = int(readln())
    for _ in range(num_cases):
        N, M = map(int, readln().split())
        maze = [list(readln()) for _ in range(N)]
        C = int(readln())

        ex, ey = [(x, y) for x in range(N) for y in range(M) if maze[x][y] == 'E'][0]
        dx, dy = [(x, y) for x in range(N) for y in range(M) if maze[x][y] == 'D'][0]
        manholes = [(x, y) for x in range(N) for y in range(M) if maze[x][y] == 'M']

        bridges = dfs_bridges_iterative(maze, dx, dy, ex, ey)
        flood_gate = random.choice(bridges)
        
        selected_covers = evaluate_manhole_cover_effectiveness(maze, N, M, manholes, dx, dy, flood_gate, C)
        #print(selected_covers)
        

        #print(flood)
        
        #for x in range(N):
        #    for y in range(M):
        #        if flood[x][y] and (x, y) in selected_covers:
        #            selected_covers.remove((x, y))
                    
        
        outln(f"{flood_gate[0][0]} {flood_gate[0][1]} {flood_gate[1][0]} {flood_gate[1][1]}")
        outln(len(selected_covers)) 
        for cover in selected_covers:
            outln(f"{cover[0]} {cover[1]}")
        path = bfs_safe_path(maze, N, M, dx, dy, ex, ey)
        outln(len(path))
        for step in path:
            outln(f"{step[0]} {step[1]}")


if __name__ == "__main__":
    main()